app/src/lib/stores/pull-request-store.ts

```diff
@@ -0,0 +1,73 @@
+import { PullRequestDatabase } from '../databases'
+import { GitHubRepository } from '../../models/github-repository'
+import { Account } from '../../models/account'
+import { API, IAPIPullRequest } from '../api'
+import { fatalError } from '../fatal-error'
+
+/** The store for GitHub Pull Requests. */
+export class PullRequestStore {
+  private db: PullRequestDatabase
+
+  public constructor(db: PullRequestDatabase) {
+    this.db = db
+  }
+
+  public async cachePullRequests(
+    repository: GitHubRepository,
+    account: Account
+  ) {
+    const api = API.fromAccount(account)
+
+    const prs = await api.fetchPullRequests(
+      repository.owner.login,
+      repository.name,
+      'open'
+    )
+
+    await this.writePullRequests(prs, repository)
+  }
+
+  public async getPullRequests(repository: GitHubRepository) {
+    const gitHubRepositoryID = repository.dbID
+    if (!gitHubRepositoryID) {
+      fatalError(
+        "Cannot get pull requests for a repository that hasn't been inserted into the database!"
+      )
+
+      return []
+    }
+
+    const pullRequests = await this.db.pullRequests
+      .where('repo_id')
+      .equals(gitHubRepositoryID)
+      .sortBy('number')
+
+    return pullRequests
+  }
+
+  private async writePullRequests(
+    pullRequests: ReadonlyArray<IAPIPullRequest>,
+    repository: GitHubRepository
+  ): Promise<void> {
+    const repoId = repository.dbID
+
+    if (!repoId) {
+      fatalError(
+        "Cannot store pull requests for a repository that hasn't been inserted into the database!"
+      )
+
+      return
+    }
+
+    const table = this.db.pullRequests
+    const apiPRs = pullRequests as Array<IAPIPullRequest>
+    const insertablePRs = apiPRs.map(x => {
+      return { repoId, ...x }
+    })
+
+    await this.db.transaction('rw', table, () => {
+      table.clear()
```

Comment:
We need to wait on this `clear` to complete so that we can be sure it's done before our `bulkAdd`. Unfortunately, because reasons, we have to use a generator function and `yield` instead of an async function and `await`.

See: https://github.com/desktop/desktop/blob/7dccd065545c92476581a509033513e8b258b546/app/src/lib/stores/issues-store.ts#L115-L117

Reply:
I saw that in documentation for deleting 1 item, but how would I use a generator with `clear` and `bulkAdd`? I was under the impression that those methods open their own transactions.

Reply:
The type inference is now working...not sure what changed, but it works without a cast now.

Reply:
You'd do something like:

```ts
await this.db.transaction('rw', table, function*() {
  yield table.clear()
  yield table.bulkAdd(insertablePRs)
})
```

>I was under the impression that those methods open their own transactions.

Maybe? It's not clear to me from the documentation. http://dexie.org/docs/Table/Table.bulkAdd().html has a warning about using it outside of a transaction which makes it non-atomic. Which makes me think it's not using a transaction.

Reply:
Would something like this work? 
```ts
    await this.db.transaction('rw', table, function*() {
      yield table
        .clear()
        .then(() => table.bulkAdd(insertablePRs))
        .catch(e => console.log('Failed to reload database', e))
    })
```

Reply:
It could, but is there a reason not to `yield` both the `clear` and `bulkAdd` as I did above? It seems clearer than `then`ing to me.

Reply:
not sure, I figured `yield` was meant to return a value and clear doesn't not. Will look into this. It looks like an `IEnumerable` but perverted by JavaScriptyness.

Reply:
In this case, `yield` is essential `await` but in a generator context instead of an async context. I dunno if that helps 😭